stl/inc/span

```diff
@@ -0,0 +1,723 @@
+        static_assert(_OtherExtent == _Extent || _OtherExtent == dynamic_extent,
+            "Mismatch between static extent and size of initializing data.");
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_OtherExtent == dynamic_extent) {
+            _STL_VERIFY(_Ext._Size() == _Extent, "Mismatch between static extent and size of initializing data");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    _NODISCARD constexpr size_t _Size() const noexcept {
+        return _Extent;
+    }
+};
+
+template <>
+class _Span_extent_type<dynamic_extent> {
+public:
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext._Size()) {}
+
+    _NODISCARD constexpr size_t _Size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+// CLASS TEMPLATE _Span_storage
+template <class _Ptr, size_t _Extent>
+class _Span_storage : public _Span_extent_type<_Extent> {
+    using _Mybase = _Span_extent_type<_Extent>;
+
+public:
+    constexpr _Span_storage() = default;
+
+    explicit constexpr _Span_storage(const size_t _Ext, const _Ptr _Data) : _Mybase(_Ext), _Mydata(_Data) {}
+
+    template <size_t _OtherExtent>
+    explicit constexpr _Span_storage(const _Span_extent_type<_OtherExtent> _Ext, const _Ptr _Data)
+        : _Mybase(_Ext), _Mydata(_Data) {}
+
+    _NODISCARD constexpr _Ptr _Data() const noexcept {
+        return _Mydata;
+    }
+
+    constexpr friend void swap(_Span_storage& _Left, _Span_storage& _Right) noexcept {
+        _STD swap(_Left._Mydata, _Right._Mydata);
+        if constexpr (_Extent == dynamic_extent) {
+            _STD swap(_Left._Mysize, _Right._Mysize);
+        }
+    }
+
+private:
+    _Ptr _Mydata{nullptr};
+};
+
+// CLASS TEMPLATE _Span_iterator
+template <class _Ptr>
+class _Span_iterator {
+public:
+#ifdef __cpp_lib_concepts
+    using iterator_concept = contiguous_iterator_tag;
+#endif // __cpp_lib_concepts
+    using iterator_category = random_access_iterator_tag;
+    using value_type        = typename iterator_traits<_Ptr>::value_type;
+    using difference_type   = typename iterator_traits<_Ptr>::difference_type;
+    using pointer           = typename iterator_traits<_Ptr>::pointer;
+    using reference         = typename iterator_traits<_Ptr>::reference;
+
+    constexpr _Span_iterator() noexcept = default;
+
+#if _ITERATOR_DEBUG_LEVEL >= 1
+    constexpr _Span_iterator(const pointer _Data, const size_t _Size, const size_t _Off) noexcept
+        : _Mydata(_Data), _Mysize(_Size), _Myoff(_Off) {}
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+    constexpr explicit _Span_iterator(const pointer _Pointer) noexcept : _Myptr(_Pointer) {}
+#endif // _ITERATOR_DEBUG_LEVEL

```

Comment:
I made it an aggregate and moved to a pure pointer implementation

Reply:
Specializations of `_Span_iterator` are currently not aggregates since it has constructors ([[dcl.init.aggr]/1.1](http://eel.is/c++draft/dcl.init.aggr#1.1)).

Reply:
Also I have to say that given the efficiency of `span` I would be horrified if people would actually use `span_iterator` in interface boundaries rather than `span`

Reply:
I don't think it can be an aggregate while meeting the null forward iterator requirements.

Reply:
"ABI boundaries" wasn't a great choice of terminology here, it suggests intentional design. I mean something more along the lines of "boundaries between clusters of inlining". Function (template)s in C++ that operate on ranges of elements frequently accept iterator+sentinel arguments and return iterators (or aggregates that contain iterators). Since it's not the case that everything always gets inlined, iterator types will appear as argument/return types in generated code. Those are the occurrences I'm concerned with: I'd like for `span` iterators to be as cheap as pointers, which means our ABI needs to see them as aggregates.

> I don't think it can be an aggregate while meeting the null forward iterator requirements.

The null forward iterator requirements apply to value-initialized iterators, not default-initialized iterators, so this isn't an issue.

Reply:
Hmmm right. I guess we go above that at this time.

Reply:
@CaseyCarter How does this work with conversions/comparisons between `iterator` and `const_iterators`, aka `_Span_iterator<const _Ty>`. There must be some way to transform `iterator` to `const_iterator`.

If we cannot convert via constructors it seems we would need to open code all the comparison operators by hand

Reply:
> If we cannot convert via constructors

We could add a `_Span_iterator<const _Ty>` conversion operator to `_Span_iterator<_Ty>` just as easily as we can add `_Span_iterator<_Ty>` converting constructor to `_Span_iterator<const _Ty>`.

Reply:
You must be able to construct a `const_iterator` from an `iterator` as a user - that's a requirement for all container iterators (which also applies to `span`).

Reply:
As far as I can tell this is fixed with the new constructor

Now I am getting a bit nervous about this being effectively a copy constructor in the case of `_OtherType == _Ty`. Does that effect the generation of other special members, e.g move?

However, given that this is a cheap aggregate the level of concern is ...

Reply:
> I believe that @CaseyCarter is referring to the Core Language notion of an aggregate

"Aggregate" is actually imprecise: I believe the criteria our ABI uses for enregistering class types is closer to aggregate + POD (trivially copyable and standard layout). This *greatly restricts* our ability to control construction.

> I am mildly concerned about people constructing iterators from raw pointers, and being an aggregate would seem to make it harder to defend against that.

Absolutely: we can't restrict construction of an aggregate by making a constructor `private` since they have no constructors. I can think of no way to do so that wouldn't make the iterator non-POD or non-aggregate.

Reply:
If span were implemented via an aggregate base then that can be used in APIs.
You also get structured binding as a freebie side-effect.
https://godbolt.org/z/N2BbZe
```
template <typename T>
struct basic_span { T* data; size_t size; };

template <typename T>
struct span : basic_span<T> { span(T* data,size_t size) : basic_span<T>{data,size} {} };

template <typename T>
auto clear(basic_span<T> bs) { while (bs.size--) *bs.data++ = T(); }
```

Reply:
@willray The discussion is about `span::iterator` not `span` itself. `span` cannot be an aggregate as the standard requires it to have certain constructors. Consequently, it cannot be an aggregate as @CaseyCarter pointed out ([[dcl.init.aggr]/1.1](http://eel.is/c++draft/dcl.init.aggr#1.1)). 

Regarding the structured bindings I believe this would actually not be the expected behaviour. A `span` denotes a range of objects. Therefore, one would expect that a structured binding would give the stored objects. However, as you proposed it it would only give the pointer to the buffer and the size, which is actually what `span` was meant to hide away.

Reply:
Summarizing the discussion:

The MS ABI likes types which are aggregate + POD (like C types); they are passed in registers. We (MSVC) want `span` and its iterators to be as efficient as possible so we can continue to encourage its use indiscriminately. 

There are a couple of downsides of making span iterators aggregates:
1. It limits our ability to control construction - aggregates cannot have constructors. In release builds, `I{pointer_value}` will be a valid expression when `I` is the type of a `span` iterator. Users will probably discover and abuse this fact, but accidental misuse should still be rare given that the implicit conversions from pointer types are forbidden and that same syntax won't work in debug builds because the aggregate will have more than one data member.

2. We won't be able to provide a guarantee that default-initialized `span` iterators wrap `null` pointer values and therefore compare equal. The standard doesn't require that behavior but MSVC has provided a stronger guarantee for our default-initialized container iterators.

I'm inclined to go the aggregate route: I think the lack of flexibility in construction is a reasonable cost for minor performance benefits when passing `span` iterators across inlining boundaries. I believe @BillyONeal agrees. @StephanTLavavej seems to be on the fence awaiting additional evidence.

Reply:
I'm ok with aggregates here; we've never gone to great lengths to prevent construction of iterators from pointers, and the default-init vs. value-init behavior is acceptable.

Reply:
I went ahead and turned `_Span_iterator` into an aggregate. I SFINAEd out the conversion for const types, as I dont believe that makes any sense?!

Reply:
@CaseyCarter currently the members of `_Span_iterator` are defaulted to `nullptr`. Wouldnt that guarantee that they compare equal to nullptr?